### PR TITLE
Issue 9967: removed codomain() function and related test from codomain.py

### PR DESCRIPTION
**Pull request by kevalds51**

### CODE REVIEW

File: sympy/calculus/tests/test_not_empty_in.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from sympy import Symbol, S, oo, sqrt
+from sympy.calculus.codomain import not_empty_in
+from sympy.sets.sets import Interval, FiniteSet, Complement, Union
+from sympy.utilities.pytest import XFAIL, raises
```
Comment (gxyd) [inline sympy/calculus/tests/test_not_empty_in.py]: This import line can be removed.

Comment (kevalds51) [inline sympy/calculus/tests/test_not_empty_in.py]: Done

**Comment by gxyd** [comment]
@aktech can you suggest some name change for the file "codomain.py"?. Also correspondingly for the test file `test_not_empty_in`. I am not sure if the name "not_empty_in" for file is appropriate.

**Comment by aktech** [comment]
@gxyd  Currently I don't recall a better name, than that. The last option could be `miscellaneous.py` but it may not help much.

**Comment by kevalds51** [comment]
how about util.py ?

**Comment by aktech** [comment]
> how about util.py ?

Sounds good.

**Comment by gxyd** [comment]
Then also change name  of file `test_not_empty_in.py` -> `test_util.py`.

**Comment by kevalds51** [comment]
on it.

**Comment by gxyd** [comment]
Looks good to me. +1. After this PR is merged, Also update the [release notes](https://github.com/sympy/sympy/wiki/Release-Notes-for-0.7.7) . Two point will have to be changed in `Major Changes` in release notes. Since i may forget that, that's why i mentioned before.

File: sympy/calculus/util.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    (-2, -1] U [2, oo)
+    """
+
+    # TODO: handle piecewise defined functions
+    # TODO: handle transcendental functions
+    # TODO: handle multivariate functions
+    if len(syms) == 0:
+        raise ValueError("A Symbol or a tuple of symbols must be given \
+                            as the third parameter")
+
```
Comment (leosartaj) [inline sympy/calculus/util.py]: You do not need backslash. You can use " instead. Python will handle spaces properly. Currently

```
In [1]: FiniteSet(x/2)
Out[1]: 
⎧x⎫
⎨─⎬
⎩2⎭

In [2]: not_empty_in(_)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-584d108ddade> in <module>()
----> 1 not_empty_in(_)

/home/leosartaj/projects/sympy/sympy/calculus/util.py in not_empty_in(finset_intersection, *syms)
     47     if len(syms) == 0:
     48         raise ValueError("A Symbol or a tuple of symbols must be given \
---> 49                             as the third parameter")
     50 
     51     if finset_intersection.is_EmptySet:

ValueError: A Symbol or a tuple of symbols must be given                             as the third parameter
```

File: sympy/calculus/util.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+from sympy import S, sympify, diff, limit, oo
```
Comment (leosartaj) [inline sympy/calculus/util.py]: sympify, diff, limit, oo are unused.

File: sympy/calculus/util.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+from sympy.calculus.singularities import singularities
```
Comment (leosartaj) [inline sympy/calculus/util.py]: This line can be removed.

File: sympy/calculus/util.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+from sympy.sets.sets import Interval, Intersection, FiniteSet, Union, Complement, Set, EmptySet
```
Comment (leosartaj) [inline sympy/calculus/util.py]: Set is not used.

Comment (leosartaj) [inline sympy/calculus/util.py]: This line exceeds 80 characters. You should put these in brackets.

File: sympy/calculus/util.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    if isinstance(finset_intersection, FiniteSet):
+        finite_set = finset_intersection
+        _sets = S.Reals
+    else:
+        finite_set = finset_intersection.args[1]
+        _sets = finset_intersection.args[0]
+
+    if not isinstance(finite_set, FiniteSet):
+        raise ValueError('A FiniteSet must be given, not %s: %s' % (type(finite_set), finite_set))
```
Comment (leosartaj) [inline sympy/calculus/util.py]: Too long.

File: sympy/calculus/util.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    if len(syms) == 0:
+        raise ValueError("A Symbol or a tuple of symbols must be given \
+                            as the third parameter")
+
+    if finset_intersection.is_EmptySet:
+        return EmptySet()
+
+    if isinstance(finset_intersection, Union):
+        elm_in_sets = finset_intersection.args[0]
+        return Union(not_empty_in(finset_intersection.args[1], *syms), elm_in_sets)
```
Comment (leosartaj) [inline sympy/calculus/util.py]: Too long.

File: sympy/calculus/util.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        finite_set = finset_intersection.args[1]
+        _sets = finset_intersection.args[0]
+
+    if not isinstance(finite_set, FiniteSet):
+        raise ValueError('A FiniteSet must be given, not %s: %s' % (type(finite_set), finite_set))
+
+    if len(syms) == 1:
+        symb = syms[0]
+    else:
+        raise NotImplementedError('more than one variables %s not handled' % (syms,))
```
Comment (leosartaj) [inline sympy/calculus/util.py]: Too long.

File: sympy/calculus/util.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    def elm_domain(expr, intrvl):
+        """ Finds the domain of an expression in any given interval """
+
+        _start = intrvl.start
+        _end = intrvl.end
+        _singularities = solveset(expr.as_numer_denom()[1], symb, domain=S.Reals)
```
Comment (leosartaj) [inline sympy/calculus/util.py]: Too long.

File: sympy/calculus/util.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        expr_domain = Complement(expr_with_sing, _singularities)
+        return expr_domain
+
+    if isinstance(_sets, Interval):
+        return Union(*[elm_domain(element, _sets) for element in finite_set])
+
+    if isinstance(_sets, Union):
+        _domain = S.EmptySet
+        for intrvl in _sets.args:
+            _domain_element = Union(*[elm_domain(element, intrvl) for element in finite_set])
```
Comment (leosartaj) [inline sympy/calculus/util.py]: Too long.

File: sympy/calculus/util.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    ValueError
+        The input is not valid.
+    RuntimeError
+        It is a bug, please report it to the github issue tracker
+        (https://github.com/sympy/sympy/issues).
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, FiniteSet, Interval, not_empty_in, sqrt, oo
```
Comment (leosartaj) [inline sympy/calculus/util.py]: Symbol, sqrt are unused.

**Comment by leosartaj** [comment]
I recommend breaking the lines that are too long in tests_util.py

**Comment by leosartaj** [comment]
Other than my few comments this looks good. +1

**Comment by leosartaj** [comment]
I do not see why 75d83080858d40511dc0842f8ad027cf29d1bbf1 is required. Also the commit message for 7758872899b2a3baa3b293303279bf0cad5325a6 seems confusing.

**Comment by kevalds51** [comment]
can anyone please restart the build job?

**Comment by leosartaj** [comment]
Done.

**Comment by kevalds51** [comment]
ping @leosartaj @gxyd

**Comment by leosartaj** [comment]
The lines in test_util are still longer than the accepted 80 characters. I recommend using a tool for this, like flake8.

File: sympy/calculus/tests/test_util.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
+from sympy import Symbol, S, oo, sqrt
+from sympy.calculus.util import not_empty_in
+from sympy.sets.sets import Interval, FiniteSet, Complement, Union
+
+
+def test_not_empty_in():
+    from sympy.abc import x
```
Comment (leosartaj) [inline sympy/calculus/tests/test_util.py]: This import should be top level.

File: sympy/calculus/util.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
+from sympy import S
+from sympy.sets.sets import (Interval, Intersection, FiniteSet, Union,
+    Complement, EmptySet)
```
Comment (aktech) [inline sympy/calculus/util.py]: This line is under indented.

Change it as:
```
from sympy.sets.sets import (Interval, Intersection, FiniteSet, Union,
                             Complement, EmptySet)
```

File: sympy/calculus/tests/test_util.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    a = Symbol('a', real=True)
```
Comment (leosartaj) [inline sympy/calculus/tests/test_util.py]: This can also be at the top.

File: sympy/calculus/util.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    >>> not_empty_in(FiniteSet(x**2/(x + 2)).intersect(Interval(1, oo)), x)
+    (-2, -1] U [2, oo)
+    """
+
+    # TODO: handle piecewise defined functions
+    # TODO: handle transcendental functions
+    # TODO: handle multivariate functions
+    if len(syms) == 0:
+        raise ValueError("A Symbol or a tuple of symbols must be given "
+                        "as the third parameter")
```
Comment (leosartaj) [inline sympy/calculus/util.py]: under indented

File: sympy/calculus/util.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+    if finset_intersection.is_EmptySet:
+        return EmptySet()
+
+    if isinstance(finset_intersection, Union):
+        elm_in_sets = finset_intersection.args[0]
+        return Union(not_empty_in(finset_intersection.args[1], *syms),
+                    elm_in_sets)
```
Comment (leosartaj) [inline sympy/calculus/util.py]: under indented

File: sympy/calculus/util.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    if isinstance(finset_intersection, FiniteSet):
+        finite_set = finset_intersection
+        _sets = S.Reals
+    else:
+        finite_set = finset_intersection.args[1]
+        _sets = finset_intersection.args[0]
+
+    if not isinstance(finite_set, FiniteSet):
+        raise ValueError('A FiniteSet must be given, not %s: %s' %
+                        (type(finite_set), finite_set))
```
Comment (leosartaj) [inline sympy/calculus/util.py]: under indented

File: sympy/calculus/util.py [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+    if len(syms) == 1:
+        symb = syms[0]
+    else:
+        raise NotImplementedError('more than one variables %s not handled' %
+                                (syms,))
```
Comment (leosartaj) [inline sympy/calculus/util.py]: under indented

File: sympy/calculus/util.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+
+    def elm_domain(expr, intrvl):
+        """ Finds the domain of an expression in any given interval """
+
+        _start = intrvl.start
+        _end = intrvl.end
+        _singularities = solveset(expr.as_numer_denom()[1], symb,
+                                    domain=S.Reals)
```
Comment (leosartaj) [inline sympy/calculus/util.py]: over-indented

**Comment by kevalds51** [comment]
ping @leosartaj

File: sympy/calculus/tests/test_util.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+                 Interval(1, -S(1)/2 + sqrt(17)/2), Interval(2, 4))
+    assert not_empty_in(FiniteSet(x/(x - 1)).intersect(S.Reals), x) == \
+        Complement(S.Reals, FiniteSet(1))
+    assert not_empty_in(FiniteSet(a/(a - 1)).intersect(S.Reals), a) == \
+        Complement(S.Reals, FiniteSet(1))
+    assert not_empty_in(FiniteSet((x**2 - 3*x + 2)/(x - 1))
+                        .intersect(S.Reals), x) == \
+        Complement(S.Reals, FiniteSet(1))
+    assert not_empty_in(FiniteSet(3, 4, x/(x - 1)).intersect(Interval(2, 3)),
+                                                             x) == \
```
Comment (leosartaj) [inline sympy/calculus/tests/test_util.py]: This very oddly broken. It's hard to read.

**Comment by leosartaj** [comment]
You can certainly write better informative commit messages in future. See [this](https://github.com/sympy/sympy/wiki/Development-workflow#id41)

**Comment by aktech** [comment]
BTW, the 80 characters rule isn't very hard & fast to follow always, sometimes we chose readability over this, although in that case the number of characters should not exceed 100 chars at max.

File: sympy/calculus/tests/test_util.py (hunk 20 above)
Comment (aktech) [inline sympy/calculus/tests/test_util.py]: For example(see https://github.com/sympy/sympy/pull/9975#issuecomment-148887124) in the above case. You could put the following in one line:

```
assert not_empty_in(FiniteSet(3, 4, x/(x - 1)).intersect(Interval(2, 3)), x) == \
```

**Comment by kevalds51** [comment]
can someone please look into the build, and why one test is not completing
Thanks

**Comment by Sumith1896** [comment]
It's a connection error. Somebody needs to restart it.

**Comment by aktech** [comment]
Build check restarted.

**Comment by kevalds51** [comment]
ping @aktech

**Comment by leosartaj** [comment]
Looks good. +1

File: sympy/calculus/tests/test_util.py [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+        Union(Interval(-sqrt(2), -1), Interval(1, 2))
+    assert not_empty_in(FiniteSet(x**2 + x, x).intersect(Interval(2, 4)), x) \
+        == Union(Interval(-sqrt(17)/2 - S(1)/2, -2),
+                 Interval(1, -S(1)/2 + sqrt(17)/2), Interval(2, 4))
+    assert not_empty_in(FiniteSet(x/(x - 1)).intersect(S.Reals), x) == \
+        Complement(S.Reals, FiniteSet(1))
+    assert not_empty_in(FiniteSet(a/(a - 1)).intersect(S.Reals), a) == \
+        Complement(S.Reals, FiniteSet(1))
+    assert not_empty_in(FiniteSet((x**2 - 3*x + 2)/(x - 1))
+                        .intersect(S.Reals), x) == \
```
Comment (aktech) [inline sympy/calculus/tests/test_util.py]: I think you can skip the 80 characters limit here, as it looks ugly as of now. Make it in one line.

File: sympy/calculus/tests/test_util.py [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+        Complement(S.Reals, FiniteSet(1))
+    assert not_empty_in(FiniteSet(3, 4, x/(x - 1)).intersect(Interval(2, 3)), x) == \
+        Union(Interval(S(3)/2, 2), FiniteSet(3))
+    assert not_empty_in(FiniteSet(x/(x**2 - 1)).intersect(S.Reals), x) == \
+        Complement(S.Reals, FiniteSet(-1, 1))
+    assert not_empty_in(FiniteSet(x, x**2)
+                        .intersect(Union(Interval(1, 3, True, True),
```
Comment (aktech) [inline sympy/calculus/tests/test_util.py]: [Same here](https://github.com/sympy/sympy/pull/9975/files#r42395812)

File: sympy/calculus/tests/test_util.py [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+from sympy.calculus.util import not_empty_in
+from sympy.sets.sets import Interval, FiniteSet, Complement, Union
+from sympy.abc import x
+
+a = Symbol('a', real=True)
+
+
+def test_not_empty_in():
+    assert not_empty_in(FiniteSet(x, 2*x).intersect(Interval(1, 2,
+                        True, False)), x) == Interval(S(1)/2, 2, True, False)
```
Comment (aktech) [inline sympy/calculus/tests/test_util.py]: This one also looks ugly.
-->
```
    assert not_empty_in(FiniteSet(x, 2*x).intersect(Interval(1, 2, True, False)), x) == \
        Interval(S(1)/2, 2, True, False)
```

File: sympy/calculus/tests/test_util.py [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+    assert not_empty_in(FiniteSet(x, x**2).intersect(Interval(1, 2)), x) == \
+        Union(Interval(-sqrt(2), -1), Interval(1, 2))
+    assert not_empty_in(FiniteSet(x**2 + x, x).intersect(Interval(2, 4)), x) \
+        == Union(Interval(-sqrt(17)/2 - S(1)/2, -2),
+                 Interval(1, -S(1)/2 + sqrt(17)/2), Interval(2, 4))
```
Comment (aktech) [inline sympy/calculus/tests/test_util.py]: -->
```
    assert not_empty_in(FiniteSet(x**2 + x, x).intersect(Interval(2, 4)), x) == \
        Union(Interval(-sqrt(17)/2 - S(1)/2, -2),
              Interval(1, -S(1)/2 + sqrt(17)/2), Interval(2, 4))
```

File: sympy/calculus/tests/test_util.py (hunk 21 above)
Comment (aktech) [inline sympy/calculus/tests/test_util.py]: -->
```
    assert not_empty_in(FiniteSet((x**2 - 3*x + 2)/(x - 1)).intersect(S.Reals), x) == \
```

File: sympy/calculus/tests/test_util.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+    assert not_empty_in(FiniteSet(x/(x**2 - 1)).intersect(S.Reals), x) == \
+        Complement(S.Reals, FiniteSet(-1, 1))
+    assert not_empty_in(FiniteSet(x, x**2)
+                        .intersect(Union(Interval(1, 3, True, True),
+                                         Interval(4, 5))), x) == \
+        Union(Interval(-sqrt(5), -2), Interval(-sqrt(3), -1, True, True),
+              Interval(1, 3, True, True), Interval(4, 5))
+    assert not_empty_in(FiniteSet(1).intersect(Interval(3, 4)), x) == S.EmptySet
+    assert not_empty_in(FiniteSet(x**2/(x + 2)).intersect(Interval(1, oo)), x) \
+        == Union(Interval(-2, -1, True, False), Interval(2, oo))
```
Comment (aktech) [inline sympy/calculus/tests/test_util.py]: Or you could also do something like this, at such situations:

```
    f_set = FiniteSet(x**2/(x + 2)).intersect(Interval(1, oo))
    assert not_empty_in(f_set, x) == Union(Interval(-2, -1, True, False),
                                           Interval(2, oo))
```

**Comment by irogerthat** [comment]
I think the commits should be squashed to one. 13 commits for a single change is way too much.

**Comment by kevalds51** [comment]
@aktech some help here would be great. I don't understand what exactly happened
I was trying to squash my commits using: git rebase -i HEAD~13

**Comment by aktech** [comment]
It looks like you messed up with rebasing, but don't worry, follow the following steps:

Step 1. checkout to this branch: `issue9967` (`$ git checkout issue9967`). Copy the following files and save it to a safe place (say Desktop).
```
sympy/calculus/__init__.py
sympy/calculus/tests/test_util.py
sympy/calculus/util.py
```
Step 2: Checkout to master branch and then Delete this branch: `issue9967` (`$ git branch -D issue9967`)

Step 3: Now create a fresh new branch named `issue9967` (`$ git checkout -b issue9967`).

Step 4: Go to `sympy.calculus` and delete `codomain.py` and `test_codomain.py` & Copy the above 3 files (from Step1) to appropriate folder.

Step 5: Now `git add .` and `$ git commit -m "Remove sympy.calculus.codomain function"`

Step 6: Now force push the changes (`$ git push origin issue9967 -f`).

**Comment by aktech** [comment]
Looks good. +1
@kevalds51  Thanks for working on this.

**Comment by gxyd** [comment]
@kevalds51 Can you also address [this comment](https://github.com/sympy/sympy/pull/9975#issuecomment-146758262)? Otherwise it would create problem later.

**Comment by gxyd** [comment]
You should modify first two points in `Major Changes`:
# TODO
- [ ] Delete the second point
- [ ] Modify the first point, change file name from `codomain.py`-> `util.py`

**Comment by gxyd** [comment]
On this page https://github.com/sympy/sympy/wiki/Release-Notes-for-1.0

**Comment by kevalds51** [comment]
@gxyd Done

**Comment by gxyd** [comment]
looks good +1.

**Pull request merged**